### ISSUE TITLE
Cards can still be inverted through non-invertible functions

### ISSUE THREAD

**Issue by phet-steele** [phetsims/function-builder#39]
@pixelzoom this is for phetsims/tasks/issues/589. Noticed on all Win 10 browsers and iPad 2 iOS 9.3.1 (Tycho).

A user can pass cards backwards through a non-invertible function by dragging the card backwards and releasing *just* before the red circle cue appears. Edit: not necessarily, read first comment.

In the case of this issue I mean what I say explicitly. The user must be **controlling** a card and moving it **backwards**, and release it at the specific moment. There must be backward movement of the card directly prior to the release. Do not release the card after dragging it forward from the input carousel, in this case you would have to drag it past the release point and "turn around" to approach the function from the right. The card does not need to be moving at the point of release. Might sound difficult to reproduce, but is actually quite easy to do (especially on accident). 
![funbu07](https://cloud.githubusercontent.com/assets/5863899/15195253/29365e8c-1784-11e6-941f-8697055033b2.gif)

Troubleshooting information (do not edit):
Name: ‪Function Builder‬
URL: http://www.colorado.edu/physics/phet/dev/html/function-builder/1.0.0-dev.25/function-builder_en.html
Version: 1.0.0-dev.25 2016-05-05 18:36:48 UTC
User Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.94 Safari/537.36
Language: en-US
Window: 2560x1315
Pixel Ratio: 1/1
WebGL: WebGL 1.0 (OpenGL ES 2.0 Chromium)
GLSL: WebGL GLSL ES 1.0 (OpenGL ES GLSL ES 1.0 Chromium)
Vendor: WebKit (WebKit WebGL)
Vertex: attribs: 16 varying: 30 uniform: 1024
Texture: size: 16384 imageUnits: 16 (vertex: 16, combined: 32)
Max viewport: 16384x16384
OES_texture_float: true
Dependencies JSON: {"assert":{"sha":"7d27130a","branch":"master"},"axon":{"sha":"5e33541f","branch":"master"},"babel":{"sha":"b461c809","branch":"master"},"brand":{"sha":"bb1cfcc6","branch":"master"},"chipper":{"sha":"184674ba","branch":"master"},"dot":{"sha":"04efe0bc","branch":"master"},"function-builder":{"sha":"3d4fe673","branch":"master"},"joist":{"sha":"48836bcf","branch":"master"},"kite":{"sha":"dce9765e","branch":"master"},"phet-core":{"sha":"c48bf320","branch":"master"},"phetcommon":{"sha":"eab7fb7f","branch":"master"},"scenery":{"sha":"7590be09","branch":"master"},"scenery-phet":{"sha":"436bf61d","branch":"master"},"sherpa":{"sha":"ab4d5c14","branch":"master"},"sun":{"sha":"9f2d40d5","branch":"master"},"tandem":{"sha":"5af5b8f8","branch":"master"},"twixt":{"sha":"53b31872","branch":"master"}}

**Comment by phet-steele** [comment]
@pixelzoom actually, there is a significant grace area of when the function may be released, this is as far out as I can go where the issue still happens:

![invert](https://cloud.githubusercontent.com/assets/5863899/15195820/cb946564-1786-11e6-852a-521ea872e776.png)

If viewed in a situation where you can see inside the function machine, the above distance is just slightly to the left of the left side of the "see inside box".

**Comment by pixelzoom** [comment]
Reproduced. Undoubtedly a problem in CardNode, where dragging is handled. Start looking in `options.endDrag`, in the `else` clause that spans lines 182-202.

**Comment by pixelzoom** [comment]
Ah... More likely in CardNode `options.translateMovable`, in the `if` statement at line 128 that begins:

```js
        // dragging to the left, check to see if blocked by a non-invertible function
        if ( dragDx < 0 ) 
```

**Comment by pixelzoom** [comment]
Related commit, to structure translateMovable similarly to endDrag: @0382612362791532964dc21d30561b8b1c3663e6

**Comment by pixelzoom** [comment]
Fixed in above commits. Labeled as 'fixed-pending-testing', will verify in a future dev version. Stay tuned...

**Comment by pixelzoom** [comment]
@phet-steele please verify in this dev version:

http://www.colorado.edu/physics/phet/dev/html/function-builder/1.0.0-dev.27/function-builder_en.html

**Comment by phet-steele** [comment]
@pixelzoom verified!

**Comment by phet-steele** [comment]
@pixelzoom, the above procedure is verified as fixed. However, in 1.0.0-dev.28 cards can still be dragged through non-invertible functions by new means. This can be done if a user drags their input/output ALL the way to the input carousel. Pinging @amyh-phet so she is aware during interviews.

![funbu08](https://cloud.githubusercontent.com/assets/5863899/15295929/427d7b94-1b50-11e6-95c5-ec3034ff91cd.gif)

**Comment by pixelzoom** [comment]
Cool, thanks for catching. And this will be an easy fix.

**Comment by pixelzoom** [comment]
Fixed in http://www.colorado.edu/physics/phet/dev/html/function-builder/1.0.0-dev.29/function-builder_en.html, please verify @phet-steele.

**Comment by phet-steele** [comment]
@pixelzoom the second procedure is indeed fixed, but now I can get cards stuck in the function machine. NO idea how, will have to look into it. This is FF.
![stuckcard](https://cloud.githubusercontent.com/assets/5863899/15366607/5a5de116-1ce2-11e6-826f-13dfe64710d0.PNG)

**Comment by pixelzoom** [comment]
@phet-steele Please investigate further. If this is independent of this issue, please create a new issue.

**Comment by phet-steele** [comment]
@pixelzoom I'm honestly at a loss. My only guess is it's tied to a split second performance drop in FF that happens soon after sim start up (can't do this in Chrome). If I'm lucky enough to release the card while simultaneously trying to invert it, and the stars align causing a performance drop at that precise moment, then maybe, just maybe, the card will get stuck. The card can't actually be inverted though, so I guess the main issue here is resolved.

**Comment by phet-steele** [comment]
I grabbed a screenshot of fps over time (recorded in the profiler). There is a quick yet noticeable drop soon after the sim finishes loading.
![funperf](https://cloud.githubusercontent.com/assets/5863899/15404819/89cc3580-1dbc-11e6-88ca-2e4456682f09.PNG)

**Comment by pixelzoom** [comment]
Are you referring to the drop at ~15 seconds?  Does this drop occur is the sim is simply idle, or while you are doing something?

**Comment by phet-steele** [comment]
Yes at 15, and I think I have to at least enter a screen for the drop. Profiler shows no drop while idle on the home screen. Also should mention this is current master.

**Comment by pixelzoom** [comment]
Moving this new issue (reported inhttps://github.com/phetsims/function-builder/issues/39#issuecomment-220080648) to #50.

Closing.
